### ISSUE TITLE
Option to pass additional parameters

### ISSUE THREAD

**Issue by AnderMuruaUria** [JuliaNLSolvers/Optim.jl#783]
It would be nice that optimize allowed the user to supply and additional argument,  (say, p), very much like in the DifferentialEquations.jl package. That p would be an additional argument to the user supplied functions (f,  g!, h!,...). This would alllow to provide constant parameters of the optimization problem in a clean way (in that case, p might be a vector of constant values). That additional argument p could be a user defined type that contains different fields with different functionalities. (For instance, a field with workspace used in the evaluation of f, g!, ...). 

Thanks for all!

**Comment by pkofod** [comment]
https://twitter.com/pkofod/status/1224074617402331142 :)

**Comment by AnderMuruaUria** [comment]
Thanks a lot for the link. 

I've recently used  Optim.optimize successfully to solve parametric minimization problems, using closures as suggested in the Section on Tips and Tricks of the documentation of Optim.jl.  

However, for relatively complex code that requires computing local minimizers in different places of the code, I prefer to be able to pass the problem parameters (and possibly preallocated memory to evaluate the objective function, its gradient, and its Hessian) explicitly. 

Fortunately, I've just realized that Julia is flexible enough to do that by myself, defining callable composite types. I briefly describe that  for eventual users of Optim.jl with similar needs:  One defines a composite type MyOptimProb with as many fields as required (for instance, real_pars and cache), and then define 
`function (optprob::MyOptimProb)(F,G,H,x)`
`               pars = optprob.real_pars`
`               cache=optprob.cache`
`               body  of function`
`end`
so that one can call Optim.optimize as
`optimize(Optim.only_fgh!(optprob!),x0)`
